### DISCUSSION TITLE
How to use pngsave's filter option (Vips::ForeignPngFilter)

### DISCUSSION THREAD

**Discussion by tonymarklove** [libvips/ruby-vips#358] [Q&A]
After digging into the `ruby-vips` code I can't see anywhere that `Vips::ForeignPngFilter` is defined.

After some investigation it appears that the PNG saving functions in Ruby accept the values as defined in the [C enum](https://github.com/libvips/libvips/blob/815b4990766325186dcbe4cf220d40db33d446ed/libvips/include/vips/foreign.h#L672) eg.

```
image.pngsave('file.png', filter: 0xF8)
```

Am I missing a simpler way to do this? Or are these filter constants missing from `ruby-vips`?

**Comment by jcupitt** [comment]
Hi @tonymarklove,

They are actually a set of flag bits your can OR together rather than an enum. I think we were just a little unsure how to best handle these in ruby.

There are quite a few missing enums -- really we should autogenerate these from introspection in the way that the python binding does.

https://github.com/libvips/pyvips/blob/master/examples/gen-enums.py

Though now I look that doesn't handle flag bit enums either.

**Reply by tonymarklove** [reply]
Hi @jcupitt 

Thanks for the information. Yes, I also noticed that they were bit flags, which I though might have been the reason that they aren't being autogenerated.

For my own purpose I have extended the `Vips` module to simply define the same constants:
```ruby
module Vips
  class ForeignPngFilter
    NONE = 0x08
    SUB = 0x10
    UP = 0x20
    AVG = 0x40
    PAETH = 0x80
    ALL = 0xF8
  end
end
```

which I can use with standard bitwise operations:

```ruby
image.pngsave('file.png', filter: (Vips::ForeignPngFilter::SUB | Vips::ForeignPngFilter::UP))
```

I agree that this is pretty uncommon to see in Ruby code though, so perhaps an option to pass one or more symbols, which are OR'd together internally, would be more idiomatic:

```
image.pngsave('file.png', filter: :all)
image.pngsave('file.png', filter: [:up, :down])
```

Assuming that sort of thing is possible with the introspection system.

In any case, thanks for your help. Since it doesn't sound like I'm missing anything obvious, I'll continue with my workaround for now.

**Reply by jcupitt** [reply]
Yes, I think `filter: [:up, :down]` should be possible. I made an enhancement issue for this.

https://github.com/libvips/ruby-vips/issues/359

Thanks for the idea!

**Comment by jcupitt** [comment]
There's a PR for this feature: 

https://github.com/libvips/ruby-vips/pull/421

So it should be officially supported soon.
